rust/kernel/regmap.rs

```diff
@@ -0,0 +1,145 @@
+// SPDX-License-Identifier: GPL-2.0
+
+use crate::error::{ptr_err_check, Error, KernelResult};
+use crate::platform_driver::{Device, PlatformDevice};
+use crate::{bindings, c_types};
+use core::mem::MaybeUninit;
+
+/// Abstraction wrapping a kernel `struct regmap`.
+///
+/// # Invariants
+///
+/// regmap locking is never disabled.
+pub struct Regmap(*mut bindings::regmap);
+
+// SAFETY: we access the underlying `struct regmap` only through its helper functions,
+// and we never disable locking by the type invariant, so:
+// - `struct regmap *` can be safely sent between threads (Send)
+// - `struct regmap`'s helper functions can be safely called from any thread (Sync)
+unsafe impl Send for Regmap {}
+unsafe impl Sync for Regmap {}
+
+impl Regmap {
+    pub fn write(&self, reg: c_types::c_uint, val: c_types::c_uint) -> KernelResult {
+        // SAFETY: FFI call.
+        // OK to coerce a shared reference to a mutable pointer, as
+        // we guarantee that a `struct regmap`'s `write` is fully synchronized.
+        let res = unsafe { bindings::regmap_write(self.0, reg, val) };
+        if res != 0 {
+            return Err(Error::from_kernel_errno(res));
+        }
+        Ok(())
+    }
+
+    pub fn read(&self, reg: c_types::c_uint) -> KernelResult<c_types::c_uint> {
+        let mut val = MaybeUninit::<c_types::c_uint>::uninit();
```

Comment:
That's a very good question. I do know that in kernel C, unnecessary initializations are definitely frowned upon. That can also be seen in `regmap`'s `read` API: it could have initialized `val` to some value on error, which could theoretically make things "safer". But they explicitly chose not to go down that path.

In Rust, we have a trade-off which doesn't have a C equivalent: do we allow unnecessary initializations if that eliminates `unsafe` code somewhere else?

Reply:
> That's a very good question. I do know that in kernel C, unnecessary initializations are definitely frowned upon. That can also be seen in `regmap`'s `read` API: it could have initialized `val` to some value on error, which could theoretically make things "safer". But they explicitly chose not to go down that path.

Note that `regmap_read()` is not initializing, just performing a store. So it leaves the question/problem to the caller.

> In Rust, we have a trade-off which doesn't have a C equivalent: do we allow unnecessary initializations if that eliminates `unsafe` code somewhere else?

It is not really about allowing unnecessary initializations, but about not adding noise where it does not matter.

Unnecessary initializations are bad mainly for performance. In cases like this one, it is a single zero word store to the stack (or maybe nothing, if we enable cross-language LTO).

Then, for C, there is also a whole discussion around diagnostics, UB and determinism. But for Rust the default is to require initialization and otherwise be very verbose about it. Plus, from what I can see, the Rust compiler does not produce diagnostics even in the most trivial case, so not even that advantage is there.

Thus the question boils down to: a possible zero-to-stack `mov` vs. code noise (one `unsafe` block, one `SAFETY` comment, one long type and one extra call) + the risk of UB now or in the future (e.g. in a refactor) + the time to review the code.

In my view, it is clear: unless something like this is in a hot path and makes a difference that can be measured, it is not worth it.

Reply:
So you are saying that it's more Rust-idiomatic to use initialization, rather than juggle `Uninit` memory around? Yes, I can certainly see the argument. Let's go with your suggestion. The kernel C people may not like the initialization when they see it, but let's cross that bridge when we come to it.

I will place a comment above the initialization to say it's not strictly required, yet Rust-idiomatic, and unlikely to matter from a performance/resource point of view. Does that sound ok?

Reply:
Well, code does not even compile if a (used) variable is not initialized, so yeah, I would say it is idiomatic! :P `MaybeUninit` is the escape hatch for cases when you really need the performance, typically in cases involving a sizable amount of memory that will be overwritten etc.

I know this can be a contentious point. But instead of looking at `unsafe` from the perspective of "*I know this is not 100% optimal -- how can I leave the code like this?!*", what we need to teach is that the right questions are "*Do I have a valid and strong enough reason to go outside the safe subset? Can I actually measure the difference? Is it worth the code noise and risk of future UB during a refactor? Have I documented my decision? Am I willing to be the one introducing a vulnerability?*, etc.

In summary: `unsafe` code needs to be justified, not the other way around. Performance can be a valid reason, of course, and here, if you think this `read()` can indeed be measurable in a hot path, go ahead and use `unsafe` -- but please add a comment why you did that!

Having said that, having an annotation on places where initialization is not needed and therefore potential candidates to be removed if shown to be needed later could be a good idea. On the other hand, someone that needs to optimize a particular bit of code would be able to look for that manually and it does not add much to the rest of the readers of the code...

Reply:
Well, (used) uninitialized variables are a hard error in Rust, so I would say it is indeed idiomatic ;) `MaybeUninit` is the escape hatch to be used where needed.

I know this can be a contentious point. But we should not approach this (or any other similar UB topic) as "*I know this is not 100% optimal -- how can I leave the code like this?!*", but rather teach that the right questions to ask oneself are:
  - *Do I have a valid and strong enough reason for introducing `unsafe`?*
  - *Can I measure the difference?*
  - *Is this in a known hot path?*
  - *Is it worth the added complexity, code noise, review time, re-review time, risk of future UB on refactor and its associated debugging time, ...?*
  - *Have I properly documented my choice of going for `unsafe`?*
  - *Am I willing to be the one introducing a vulnerability?*
  - *...*

In summary, the rule should be: *avoid `unsafe` unless justified*. Performance is a valid justification, of course, and if you think this `read()` will end up in a hot path, go ahead and use `unsafe` -- but please add a comment explaining that!

An annotation where we perform initializations that could be potentially removed if needed later on might be a good idea. On the other hand, someone optimizing a particular code path can look manually for that...

We can discuss this in the call later today. I will also add this into the coding guidelines I am writing for `Doc/`.

Reply:
Hmm... Somehow the reply got lost in a refresh I did and I had to retype it again from memory. And when I submitted the new one, the other one did not show either.

Oh well, deleted the previous one. Sorry for the noise.